python/cudf/dataframe/dataframe.py

```diff
@@ -1499,169 +1484,158 @@ def fix_name(name, suffix):
+            left_of_on = []
+            for name in lhs._cols.keys():
+                if name not in left_on:
+                    for i in range(len(gdf_result)):
+                        if gdf_result[i][2] == name:
+                            left_of_on.append(gdf_result.pop(i))
+                            break
+            in_on = []
+            for name in itertools.chain(lhs._cols.keys(), rhs._cols.keys()):
+                if name in left_on or name in right_on:
+                    for i in range(len(gdf_result)):
+                        if gdf_result[i][2] == name:
+                            in_on.append(gdf_result.pop(i))
+                            break
+            right_of_on = []
+            for name in rhs._cols.keys():
+                if name not in right_on:
+                    for i in range(len(gdf_result)):
+                        if gdf_result[i][2] == name:
+                            right_of_on.append(gdf_result.pop(i))
+                            break
+            result = sorted(left_of_on, key=lambda x: str(x[2])) + \
+                sorted(in_on, key=lambda x: str(x[2])) + \
+                sorted(right_of_on, key=lambda x: str(x[2]))
+        else:
+            for org_name in org_names:
+                for i in range(len(gdf_result)):
+                    if gdf_result[i][2] == org_name:
+                        result.append(gdf_result.pop(i))
+                        break
+            assert(len(gdf_result) == 0)
+
+        # Build a new data frame based on the merged columns from GDF
+        df = DataFrame()
+        for col, valid, name in result:
+            if isinstance(col, nvstrings.nvstrings):
+                df[name] = col
+            else:
+                mask = None
+                if valid is not None:
+                    mask = Buffer(valid)
+                df[name] = columnops.build_column(
+                    Buffer(col),
+                    dtype=col.dtype,
+                    mask=mask,
+                    categories=col_with_categories.get(name, None),
+                )
 
+        # Let's make the "index as column" back into an index
         if left_index and right_index:
-            df = df.set_index(lhs.LEFT_RIGHT_INDEX_NAME)
-            df = df.sort_values(df.columns[0])
-            df.index.name = self._original_left_index_name
-        elif right_index and left_on:
-            new_index = Series(lhs.index,
-                               index=RangeIndex(0, len(lhs[left_on[0]])))
-            indexed = lhs[left_on[0]][df[left_on[0]]-1]
-            new_index = new_index[indexed-1]
-            df.index = new_index
-        elif left_index and right_on:
-            new_index = Series(rhs.index,
-                               index=RangeIndex(0, len(rhs[right_on[0]])))
-            indexed = rhs[right_on[0]][df[right_on[0]]-1]
-            new_index = new_index[indexed-1]
-            df.index = new_index
+            df._index = df[merge_index_name]
```

Comment:
We have an index setter method here so you should be able to use it here instead of modifying the internal attribute yourself.